src/hotspot/cpu/x86/stubGenerator_x86_32.cpp

```diff
@@ -3288,6 +3288,491 @@ class StubGenerator: public StubCodeGenerator {
+    __ vpaddd(ymm_dVec, ymm_dVec, ymm_scratch, Assembler::AVX_256bit);
+    __ vpaddd(ymm_dVec, ymm_dVec, ExternalAddress(StubRoutines::x86::chacha20_counter_addmask_avx2()), Assembler::AVX_256bit, rax);
+
+    // Write the data to the keystream array
+    // Each half of the YMM has to be written 64 bytes apart from
+    // each other in memory so the final keystream buffer holds
+    // two consecutive keystream blocks.
+    __ vextracti128(Address(result, 0), ymm_aVec, 0);
+    __ vextracti128(Address(result, 64), ymm_aVec, 1);
+    __ vextracti128(Address(result, 16), ymm_bVec, 0);
+    __ vextracti128(Address(result, 80), ymm_bVec, 1);
+    __ vextracti128(Address(result, 32), ymm_cVec, 0);
+    __ vextracti128(Address(result, 96), ymm_cVec, 1);
+    __ vextracti128(Address(result, 48), ymm_dVec, 0);
+    __ vextracti128(Address(result, 112), ymm_dVec, 1);
+
+    // This function will always write 128 bytes into the key stream buffer
+    // and that length should be returned through %rax.
+    __ movl(rax, 128);
+
+    __ leave();
+    __ ret(0);
+    return start;
+  }
+
+  /* Add mask for 4-block ChaCha20 Block calculations */
+  address chacha20_ctradd_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20_ctradd_avx512");
+    address start = __ pc();
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000001, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000002, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000003, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    __ emit_data(0x00000000, relocInfo::none, 0);
+    return start;
+  }
+
+  /* The 4-block AVX512-enabled ChaCha20 block function implementation */
+  address generate_chacha20Block_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20Block");
+    address start = __ pc();
+
+    Label L_twoRounds;
+    const Register state        = rbx;
+    const Register result       = rdx;
+    const Register loopCounter  = rcx;
+
+    const XMMRegister zmm_aVec = xmm0;
+    const XMMRegister zmm_bVec = xmm1;
+    const XMMRegister zmm_cVec = xmm2;
+    const XMMRegister zmm_dVec = xmm3;
+    const XMMRegister zmm_aState = xmm4;
+    const XMMRegister zmm_bState = xmm5;
+    const XMMRegister zmm_cState = xmm6;
+    const XMMRegister zmm_dState = xmm7;
+    const XMMRegister zmm_addMask = xmm8;
```

Comment:
Whoops!  It looks like there may not be an xmm8 register available to 32-bit architectures.  This may need a little creative restructuring in order to make it work.